### ISSUE TITLE
Support Autodiscovery of search plugins / OpenSearch description format

### ISSUE THREAD

**Issue by return42** [mozilla-mobile/fenix#21883]
### What is the user problem or growth opportunity you want to see solved?

[Autodiscovery of search plugins](https://developer.mozilla.org/en-US/docs/Web/OpenSearch#autodiscovery_of_search_plugins) gives the user the ability to easily add alternative (and privacy-respecting) search engines, including autocompletion (and other) features.

Common desktop browsers (e.g. [Firefox](https://support.mozilla.org/en-US/kb/add-or-remove-search-engine-firefox),  [Chromium](https://www.chromium.org/tab-to-search) or [MS-Edge](https://support.microsoft.com/en-us/microsoft-edge/change-your-default-search-engine-f863c519-5994-a8ed-6859-00fbc123b782)) offer [Autodiscovery of search plugins](https://developer.mozilla.org/en-US/docs/Web/OpenSearch#autodiscovery_of_search_plugins) based on the [OpenSearch protocol](https://github.com/dewitt/opensearch).  But it is very hard to find a mobile browser that can do that too (_unique selling point?_).

### How do you know that this problem exists today? Why is this important?

On the part of the users ... 

Today there is only the ability to add additional search engines manually, using the `%s` notation, but this is not comparable to search engines that are built in. E.g. those manually additions do not support autocompletion / what matters most from the users point of view.

On the part of the search engine providers ...

Non commercial search engines, e.g  I am one of the maintainer of [SearXNG](https://github.com/searxng/searxng) (a metasearch engine), we do support the [OpenSearchDescription](https://developer.mozilla.org/en-US/docs/Web/OpenSearch) in our [templating](https://github.com/searxng/searxng/blob/master/searx/templates/__common__/opensearch.xml) but I don't know which mobile browser I could recommend to our user base. Personally I like to use Firefox for Android, but the missing of [Autodiscovery of search plugins](https://developer.mozilla.org/en-US/docs/Web/OpenSearch#autodiscovery_of_search_plugins) makes me hesitate to recommend Firefox for Android to the [SearXNG](https://github.com/searxng/searxng)  user base.

### Who will benefit from it?

The whole Firefox for Android user base could add search engines they trust in.

### UX

Alternatives that comes in mind ..

- [Add search engine by long-press in search text entry box](https://github.com/mozilla-mobile/fenix/issues/6315)
- [Add a search engine from the Search bar](https://support.mozilla.org/en-US/kb/add-or-remove-search-engine-firefox#w_add-a-search-engine-from-the-search-bar)

### Additional related comments & issues

- https://github.com/mozilla-mobile/fenix/issues/8889
- https://github.com/mozilla-mobile/fenix/issues/6315#issuecomment-941054722
- https://github.com/mozilla-mobile/fenix/issues/5577
- https://github.com/mozilla-mobile/fenix/issues/14731
- https://github.com/mozilla-mobile/fenix/issues/6753
- https://github.com/mozilla-mobile/fenix/issues/15915

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-18280)

**Comment by kbrosnan** [comment]
If we want autodiscovery to work it would require #8889 first and could be part of the initial work. For enhancements that have not been done a general issue is enough. #6615 is a different mechanism than opensearch, looking at the query action and pre building the `%s` search URL.

**Comment by d-g** [comment]
> Common desktop browsers (e.g. Firefox, Chromium or MS-Edge) offer Autodiscovery of search plugins based on the OpenSearch protocol. But it is very hard to find a mobile browser that can do that too (unique selling point?).

Are you joking?  Chromium-like browsers are hard to find?  They often come preinstalled on mobile machines.

> I am one of the maintainer of SearXNG

Happy to meet you, by the way. :-)

> but I don't know which mobile browser I could recommend to our user base.

While the way the UI is designed indeed prevents user to add some popular search engines, it works absolutely fine with Searx and Searx-NG.

If you are really unaware, how it is supposed to be used:

1. You navigate to, say, https://searx.be/ (to the _root_ page).
2. Perform any search.
3. Go to Settings / Search Engine / Recently Visited and now you can make it a default search.

Alas, there is no any other search but default (i. e. no icons to choose an alternative one, no keywords), but that was not your issue.

> Personally I like to use Firefox for Android

Since your are fine with recommending nonfree software, Google Chrome is an obvious choice.

As for anyone who respects his freedom, I would suggest him to look at https://bromite.org, where they make (a) vanilla Chromium builds and (b) their custom flavour, that adds some must-have features (such as .user.js support).

**Comment by return42** [comment]
> Are you joking? Chromium-like browsers are hard to find? They often come preinstalled on mobile machines.

I'm aware of Chromium-based browsers on android devices, but they are full of bloatware and tracking features. This is the reason why I am asking here at _Firefox for Android_.

**Comment by vityav** [comment]
> While the way the UI is designed indeed prevents user to add some popular search engines, it works absolutely fine with Searx and Searx-NG.
> 
> If you are really unaware, how it is supposed to be used:
> 
>     1. You navigate to, say, https://searx.be/ (to the _root_ page).
> 
>     2. Perform any search.
> 
>     3. Go to Settings / Search Engine / Recently Visited and now you can make it a default search.
> 

To be clear, you must be using some sort of Chromium browser. I've just attempted this in Firefox 94.1.2, Firefox Beta 95.0.0-beta.5, and Firefox Nightly 96.0a1, and none have a "Recently Visited" search engine listing after searches on searx.be and arxiv.org. Currently the only solution in Firefox for Android still seems to be manually adding search engines via GET url, which as mentioned means losing features such as autocomplete.

Also, I installed and tested Chrome, Vivaldi, Opera, Brave, Edge, Kiwi, DuckDuckGo, and even the advertised Bromite browser. The only ones with the ability to add custom search engines from OpenSearch are the ones closest to stock Chromium (Chrome, Kiwi, and Bromite), so the point about this being a unique feature is still fairly applicable, and the ability to implement it with a more intuitive UI would be novel (i.e. don't need to learn how to do it from a github issue tracker).
